### ISSUE TITLE
should Instance.exports be a getter on Instance.prototype instead of data property?

### ISSUE THREAD

**Issue by lukewagner** [WebAssembly/design#1007]
Scanning the JS spec, it seems pretty irregular to have the `exports` property of instances be defined as an own data property directly on the `Instance` objects as opposed to a getter on `Instance.prototype` that reads an internal [[Exports]] property of `Instance` objects.  I'd be pretty surprised if making this change broke any code, so it seems like a minor tweak we could make after MVP shipping.

I also have a vague impression that having an internal immutable [[Exports]] property might be useful in the future if any JS API needed to find the exports object of an instance.

**Comment by rossberg-chromium** [comment]
I thought the idea of the current design is to mirror a module namespace object. That also carries the exports as own properties (https://tc39.github.io/ecma262/#sec-module-namespace-exotic-objects).

**Comment by jfbastien** [comment]
> I thought the idea of the current design is to mirror a module namespace object. That also carries the exports as own properties (https://tc39.github.io/ecma262/#sec-module-namespace-exotic-objects).

That's my understanding as well, and is exactly what we've implemented in JSC. On instance creation we literally do:

```cpp
putDirect(vm, Identifier::fromString(&vm, "exports"), moduleNamespaceObject, None);
```

I thought we'd try to move *closer* to JS modules, not further.

**Comment by lukewagner** [comment]
@rossberg-chromium I think you're referring to the exports object; I'm talking about the `exports` *property* on `Instance` objects.  I didn't know there was an analogue to JS for `exports` property, but certainly happy to follow it if there is.

**Comment by jfbastien** [comment]
@lukewagner do you think this can move forward with a PR? (and can you do that PR :wink:)

**Comment by domenic** [comment]
Note that data properties are not possible in Web IDL, so if there is an idea to expose these APIs to wasm eventually via some Web IDL binding, the data property design makes it impossible for wasm to see `Instance` objects.

**Comment by lukewagner** [comment]
Thanks @domenic!  That sounds like the strong reason to make it a getter that I was lacking.  I'm assuming the symmetry-with-module-objects concern is addressed by my previous comment, so I'll write up a PR in a bit.

**Comment by rossberg-chromium** [comment]
@lukewagner, yes, thanks for clarifying.

**Comment by lukewagner** [comment]
Resolved by #1062.
